scripts/build/check_init_priorities.py

```diff
@@ -0,0 +1,319 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Checks the initialization priorities
+
+This script parses the object files in the specified build directory, creates a
+list of known devices and their effective initialization priorities and
+compares that with the device dependencies inferred from the devicetree
+hierarchy.
+
+This can be used to detect devices that are initialized in the incorrect order,
+but also devices that are initialized at the same priority but depends on each
+other, which can potentially break if the linking order is changed.
+"""
+
+import argparse
+import logging
+import os
+import pathlib
+import pickle
+import sys
+
+from elftools.elf.elffile import ELFFile
+from elftools.elf.relocation import RelocationSection
+from elftools.elf.sections import SymbolTableSection
+
+# This is needed to load edt.pickle files.
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..",
+                                "dts", "python-devicetree", "src"))
+from devicetree import edtlib  # pylint: disable=unused-import
+
+# Prefix used for relocation sections containing initialization data, as in
+# sequence of "struct init_entry".
+_INIT_SECTION_PREFIX = (".rel.z_init_", ".rela.z_init_")
+
+# Prefix used for "struct device" reference initialized based on devicetree
+# entries with a known ordinal.
+_DEVICE_ORD_PREFIX = "__device_dts_ord_"
+
+# File name suffix for object files to be scanned.
+_OBJ_FILE_SUFFIX = ".c.obj"
+
+# Defined init level in order of priority.
+_DEVICE_INIT_LEVELS = ["EARLY", "PRE_KERNEL_1", "PRE_KERNEL_2", "POST_KERNEL",
+                      "APPLICATION", "SMP"]
+
+# Build project settings file, used to detect and skip sub-builds.
+_SETTINGS_FILE = "zephyr_settings.txt"
+
+# List of compatibles for node where the initialization priority should be the
+# opposite of the device tree inferred dependency.
+_INVERTED_PRIORITY_COMPATIBLES = set([
```

Comment:
I believe we should probably incentivize changing the devicetree nodes to comply with the init order in the future:

Changing around the devicetree a bit from this:
```
gmac: gmac {
	mdio: mdio {
	};
	phy: phy {
	};
};
```
to:
```
gmac: gmac {
    phy = <&phy>;
};
phy: phy {
    mdio = <&mdio>;
};
mdio: mdio {
};
```
should comply with this script. This way, we don't have to maintain this list of exceptions in the future.

Reply:
That's certainly an option, certainly something that could be cool do in the same way on the various network driver. Let me remove that override so that the specific platform fails again, will still leave the code to invert the check in there since I suspect we'll use it for some case anyway. Then once the tool is merged we can open the conversation with the driver owners.

Reply:
Dropped